src/main/java/com/minecolonies/coremod/util/InventoryUtils.java

```diff
@@ -294,603 +363,1248 @@ public static boolean hasToolLevel(final String tool, final ItemStack stack, fin
+                                                                        @NotNull final Predicate<ItemStack> itemStackSelectionPredicate)
+    {
+        if (!provider.hasCapability(ITEM_HANDLER_CAPABILITY, facing))
+        {
+            return Collections.EMPTY_LIST;
+        }
+
+        return filterItemHandler(provider.getCapability(ITEM_HANDLER_CAPABILITY, facing), itemStackSelectionPredicate);
+    }
+
+    /**
+     * Returns the index of the first occurrence of the block in the {@link
+     * ICapabilityProvider} for a given {@link EnumFacing}.
+     *
+     * @param provider   {@link ICapabilityProvider} to check.
+     * @param facing     The facing to check for.
+     * @param block      Block to find.
+     * @param itemDamage the damage value.
+     * @return Index of the first occurrence.
+     */
+    public static int findFirstSlotInProviderForSideWith(@NotNull final ICapabilityProvider provider, @Nullable EnumFacing facing, @NotNull final Block block, int itemDamage)
+    {
+        return findFirstSlotInProviderForSideWith(provider, facing, getItemFromBlock(block), itemDamage);
+    }
+
+    /**
+     * Returns the index of the first occurrence of the Item with the given
+     * ItemDamage in the {@link ICapabilityProvider} for a given {@link
+     * EnumFacing}.
+     *
+     * @param provider   {@link ICapabilityProvider} to check
+     * @param facing     The facing to check for.
+     * @param targetItem Item to find.
+     * @param itemDamage The damage value to match on the stack.
+     * @return Index of the first occurrence
+     */
+    public static int findFirstSlotInProviderForSideWith(@NotNull final ICapabilityProvider provider, @Nullable EnumFacing facing, @NotNull final Item targetItem, int itemDamage)
+    {
+        return findFirstSlotInProviderForSideWith(provider, facing, (ItemStack stack) -> compareItems(stack, targetItem, itemDamage));
+    }
+
+    /**
+     * Returns the index of the first occurrence of an ItemStack that matches
+     * the given predicate in the {@link ICapabilityProvider} for a given {@link
+     * EnumFacing}.
+     *
+     * @param provider                    Provider to check
+     * @param facing                      The facing to check for.
+     * @param itemStackSelectionPredicate The predicate to match.
+     * @return Index of the first occurrence
+     */
+    public static int findFirstSlotInProviderForSideWith(
+                                                          @NotNull final ICapabilityProvider provider,
+                                                          @Nullable EnumFacing facing,
+                                                          @NotNull final Predicate<ItemStack> itemStackSelectionPredicate)
+    {
+        if (!provider.hasCapability(ITEM_HANDLER_CAPABILITY, facing))
+        {
+            return -1;
+            //TODO: Later harden contract to remove compare on slot := -1
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)